OpenRA.Mods.Cnc/Traits/MadTank.cs

```diff
@@ -215,15 +147,114 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 					self.CancelActivity();
 
 				self.SetTargetLine(order.Target, Color.Red);
-				self.QueueActivity(new MoveAdjacentTo(self, order.Target, targetLineColor: Color.Red));
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+				self.QueueActivity(new DetonationSequence(self, this, order.Target));
 			}
 			else if (order.OrderString == "Detonate")
 			{
 				if (!order.Queued)
 					self.CancelActivity();
 
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+				self.QueueActivity(new DetonationSequence(self, this));
+			}
+		}
+
+		class DetonationSequence : Activity
+		{
+			readonly Actor self;
+			readonly MadTank mad;
+			readonly WithFacingSpriteBody wfsb;
+			readonly ScreenShaker screenShaker;
+			readonly bool assignTargetOnFirstRun;
+
+			int ticks;
+			bool initiated;
+			Target target;
+
+			public DetonationSequence(Actor self, MadTank mad)
+				: this(self, mad, Target.Invalid)
+			{
+				assignTargetOnFirstRun = true;
+			}
+
+			public DetonationSequence(Actor self, MadTank mad, Target target)
+			{
+				this.self = self;
+				this.mad = mad;
+				this.target = target;
+				wfsb = self.Trait<WithFacingSpriteBody>();
+				screenShaker = self.World.WorldActor.Trait<ScreenShaker>();
+			}
+
+			protected override void OnFirstRun(Actor self)
+			{
+				if (assignTargetOnFirstRun)
+					target = Target.FromCell(self.World, self.Location);
+				else
+					QueueChild(new MoveAdjacentTo(self, target, targetLineColor: Color.Red));
+			}
+
+			public override bool Tick(Actor self)
+			{
+				if (!initiated)
+				{
+					// If the target has died while we were moving, we should abort detonation.
+					if (target.Type == TargetType.Invalid)
+						return true;
+
+					if (mad.conditionManager != null && !string.IsNullOrEmpty(mad.info.DeployedCondition))
+						mad.conditionManager.GrantCondition(self, mad.info.DeployedCondition);
+
+					self.World.AddFrameEndTask(w => EjectDriver());
+					if (mad.info.ThumpSequence != null)
+						wfsb.PlayCustomAnimationRepeating(self, mad.info.ThumpSequence);
+
+					IsInterruptible = false;
+					initiated = true;
+				}
+
+				if (++ticks % mad.info.ThumpInterval == 0)
+				{
+					if (mad.info.ThumpDamageWeapon != null)
+					{
+						// Use .FromPos since this weapon needs to affect more than just the MadTank actor
+						mad.info.ThumpDamageWeaponInfo.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
+					}
+
+					screenShaker.AddEffect(mad.info.ThumpShakeTime, self.CenterPosition, mad.info.ThumpShakeIntensity, mad.info.ThumpShakeMultiplier);
+				}
+
+				if (ticks == mad.info.ChargeDelay)
+					Game.Sound.Play(SoundType.World, mad.info.ChargeSound, self.CenterPosition);
+
+				return ticks == mad.info.DetonationDelay;
+			}
+
+			protected override void OnLastRun(Actor self)
+			{
+				Game.Sound.Play(SoundType.World, mad.info.DetonationSound, self.CenterPosition);
+
+				self.World.AddFrameEndTask(w =>
```

Comment:
This detonates the tank immediately, without charging up, if `target.Type == TargetType.Invalid`!

Repro case: target an enemy building, then destroy it using other units before the mad tank reaches it.

Reply:
Fixed.